File: repertoire/data-integration/providers/capture_mode/WebBookmark.swift
// Data Integration Kit - Web Bookmark Capture Provider
// Lightweight metadata-only capture via OpenGraph, Twitter Card, and HTML meta tags

import Foundation

public final class WebBookmarkCaptureProvider {
    private static let maxHeadBytes = 16384

    public init() {}

    public func capture(input: CaptureInput, config: CaptureConfig) async throws -> CaptureItem {
        guard let urlString = input.url, let url = URL(string: urlString) else {
            throw CaptureError.missingURL
        }

        var request = URLRequest(url: url)
        request.setValue("bytes=0-\(Self.maxHeadBytes)", forHTTPHeaderField: "Range")
        let (data, _) = try await URLSession.shared.data(for: request)
        guard let html = String(data: data, encoding: .utf8) else {
            throw CaptureError.parseError("Unable to decode HTML as UTF-8")
        }

        let og = extractOpenGraph(html: html)
        let twitter = extractTwitterCard(html: html)
        let htmlMeta = extractHtmlMeta(html: html, baseUrl: urlString)

        let bookmark = mergeBookmarkData(og, twitter, htmlMeta)
        let title = bookmark["title"] ?? "Untitled Bookmark"

        var contentParts = ["# \(title)"]
        if let desc = bookmark["description"] { contentParts.append(desc) }
        if let site = bookmark["siteName"] { contentParts.append("Site: \(site)") }
        if let author = bookmark["author"] { contentParts.append("Author: \(author)") }
        if let image = bookmark["image"] { contentParts.append("Image: \(image)") }
        if let favicon = bookmark["favicon"] { contentParts.append("Favicon: \(favicon)") }

        return CaptureItem(
            content: contentParts.joined(separator: "\n"),
            sourceMetadata: SourceMetadata(
                title: title,
                url: bookmark["canonicalUrl"] ?? urlString,
                capturedAt: ISO8601DateFormatter().string(from: Date()),
                contentType: "application/x-bookmark",
                author: bookmark["author"],
                tags: ["bookmark", bookmark["type"] ?? "webpage"],
                source: "web_bookmark"
            ),
            rawData: (config.options?["includeRawMeta"] as? Bool == true) ? bookmark : nil
        )
    }

    public func supports(input: CaptureInput) -> Bool {
        guard let url = input.url else { return false }
        return url.hasPrefix("http://") || url.hasPrefix("https://")
    }

    private func extractMetaTag(html: String, property: String) -> String? {
        let escaped = NSRegularExpression.escapedPattern(for: property)
        let patterns = [
            "(?i)<meta[^>]+(?:property|name)=[\"']\(escaped)[\"'][^>]+content=[\"']([^\"']+)[\"']",
            "(?i)<meta[^>]+content=[\"']([^\"']+)[\"'][^>]+(?:property|name)=[\"']\(escaped)[\"']"
        ]
        for pattern in patterns {
            if let regex = try? NSRegularExpression(pattern: pattern),
               let match = regex.firstMatch(in: html, range: NSRange(html.startIndex..., in: html)),
               let range = Range(match.range(at: 1), in: html) {
                let val = String(html[range]).trimmingCharacters(in: .whitespaces)
                if !val.isEmpty { return val }
            }
        }
        return nil
    }

    private func extractOpenGraph(html: String) -> [String: String] {
        var result: [String: String] = [:]
        if let v = extractMetaTag(html: html, property: "og:title") { result["title"] = v }
        if let v = extractMetaTag(html: html, property: "og:description") { result["description"] = v }
        if let v = extractMetaTag(html: html, property: "og:image") { result["image"] = v }
        if let v = extractMetaTag(html: html, property: "og:site_name") { result["siteName"] = v }
        if let v = extractMetaTag(html: html, property: "og:type") { result["type"] = v }
        if let v = extractMetaTag(html: html, property: "og:url") { result["canonicalUrl"] = v }
        return result
    }

    private func extractTwitterCard(html: String) -> [String: String] {
        var result: [String: String] = [:]
        if let v = extractMetaTag(html: html, property: "twitter:card") { result["twitterCard"] = v }
        if let v = extractMetaTag(html: html, property: "twitter:title") { result["title"] = v }
        if let v = extractMetaTag(html: html, property: "twitter:description") { result["description"] = v }
        if let v = extractMetaTag(html: html, property: "twitter:image") { result["image"] = v }
        return result
    }

    private func extractHtmlMeta(html: String, baseUrl: String) -> [String: String] {
        var result: [String: String] = [:]
        if let regex = try? NSRegularExpression(pattern: "(?i)<title>([^<]+)</title>"),
           let match = regex.firstMatch(in: html, range: NSRange(html.startIndex..., in: html)),
           let range = Range(match.range(at: 1), in: html) {
            result["title"] = String(html[range]).trimmingCharacters(in: .whitespaces)
        }
        if let v = extractMetaTag(html: html, property: "description") { result["description"] = v }
        if let v = extractMetaTag(html: html, property: "author") { result["author"] = v }
        result["favicon"] = extractFavicon(html: html, baseUrl: baseUrl)
        if let regex = try? NSRegularExpression(pattern: #"(?i)<link[^>]+rel=["']canonical["'][^>]+href=["']([^"']+)["']"#),
           let match = regex.firstMatch(in: html, range: NSRange(html.startIndex..., in: html)),
           let range = Range(match.range(at: 1), in: html) {
            result["canonicalUrl"] = String(html[range]).trimmingCharacters(in: .whitespaces)
        }
        return result
    }

    private func extractFavicon(html: String, baseUrl: String) -> String {
        let patterns = [
            #"(?i)<link[^>]+rel=["'](?:shortcut )?icon["'][^>]+href=["']([^"']+)["']"#,
            #"(?i)<link[^>]+href=["']([^"']+)["'][^>]+rel=["'](?:shortcut )?icon["']"#
        ]
        for pattern in patterns {
            if let regex = try? NSRegularExpression(pattern: pattern),
               let match = regex.firstMatch(in: html, range: NSRange(html.startIndex..., in: html)),
               let range = Range(match.range(at: 1), in: html) {
                let href = String(html[range])
                if let resolved = URL(string: href, relativeTo: URL(string: baseUrl)) {
                    return resolved.absoluteString
                }
                return href
            }
        }
        if let base = URL(string: baseUrl), let favicon = URL(string: "/favicon.ico", relativeTo: base) {
            return favicon.absoluteString
        }
        return "\(baseUrl)/favicon.ico"
    }

    private func mergeBookmarkData(_ sources: [String: String]...) -> [String: String] {
        var result: [String: String] = [:]
        for source in sources {
            for (key, value) in source {
                if result[key] == nil { result[key] = value }
            }
        }
        return result
    }
}

File: repertoire/data-integration/providers/capture_mode/web_bookmark.ts
// Data Integration Kit - Web Bookmark Capture Provider
// Lightweight metadata-only capture via OpenGraph, Twitter Card, and HTML meta tags

export const PROVIDER_ID = 'web_bookmark';
export const PLUGIN_TYPE = 'capture_mode';

export interface CaptureInput {
  url?: string;
  file?: Buffer;
  email?: string;
  shareData?: unknown;
}

export interface CaptureConfig {
  mode: string;
  options?: Record<string, unknown>;
}

export interface SourceMetadata {
  title: string;
  url?: string;
  capturedAt: string;
  contentType: string;
  author?: string;
  tags?: string[];
  source?: string;
}

export interface CaptureItem {
  content: string;
  sourceMetadata: SourceMetadata;
  rawData?: unknown;
}

interface BookmarkData {
  title?: string;
  description?: string;
  image?: string;
  favicon?: string;
  canonicalUrl?: string;
  siteName?: string;
  twitterCard?: string;
  twitterSite?: string;
  author?: string;
  type?: string;
}

const MAX_HEAD_BYTES = 16384; // Only fetch first 16KB for metadata extraction

function extractMetaTag(html: string, property: string): string | undefined {
  const patterns = [
    new RegExp(`<meta[^>]+(?:property|name)=["']${property}["'][^>]+content=["']([^"']+)["']`, 'i'),
    new RegExp(`<meta[^>]+content=["']([^"']+)["'][^>]+(?:property|name)=["']${property}["']`, 'i'),
  ];
  for (const pattern of patterns) {
    const match = html.match(pattern);
    if (match?.[1]) return match[1].trim();
  }
  return undefined;
}

function extractOpenGraph(html: string): Partial<BookmarkData> {
  return {
    title: extractMetaTag(html, 'og:title'),
    description: extractMetaTag(html, 'og:description'),
    image: extractMetaTag(html, 'og:image'),
    siteName: extractMetaTag(html, 'og:site_name'),
    type: extractMetaTag(html, 'og:type'),
    canonicalUrl: extractMetaTag(html, 'og:url'),
  };
}

function extractTwitterCard(html: string): Partial<BookmarkData> {
  return {
    twitterCard: extractMetaTag(html, 'twitter:card'),
    title: extractMetaTag(html, 'twitter:title'),
    description: extractMetaTag(html, 'twitter:description'),
    image: extractMetaTag(html, 'twitter:image'),
    twitterSite: extractMetaTag(html, 'twitter:site'),
  };
}

function extractHtmlTitle(html: string): string | undefined {
  const match = html.match(/<title>([^<]+)<\/title>/i);
  return match?.[1]?.trim();
}

function extractMetaDescription(html: string): string | undefined {
  return extractMetaTag(html, 'description');
}

function extractFavicon(html: string, baseUrl: string): string | undefined {
  const patterns = [
    /<link[^>]+rel=["'](?:shortcut )?icon["'][^>]+href=["']([^"']+)["']/i,
    /<link[^>]+href=["']([^"']+)["'][^>]+rel=["'](?:shortcut )?icon["']/i,
    /<link[^>]+rel=["']apple-touch-icon["'][^>]+href=["']([^"']+)["']/i,
  ];
  for (const pattern of patterns) {
    const match = html.match(pattern);
    if (match?.[1]) {
      try { return new URL(match[1], baseUrl).href; } catch { return match[1]; }
    }
  }
  try { return new URL('/favicon.ico', baseUrl).href; } catch { return undefined; }
}

function extractCanonicalUrl(html: string): string | undefined {
  const match = html.match(/<link[^>]+rel=["']canonical["'][^>]+href=["']([^"']+)["']/i);
  return match?.[1]?.trim();
}

function extractAuthor(html: string): string | undefined {
  return extractMetaTag(html, 'author') ?? extractMetaTag(html, 'article:author');
}

function mergeBookmarkData(...sources: Partial<BookmarkData>[]): BookmarkData {
  const result: BookmarkData = {};
  for (const source of sources) {
    for (const [key, value] of Object.entries(source)) {
      if (value && !(result as Record<string, unknown>)[key]) {
        (result as Record<string, unknown>)[key] = value;
      }
    }
  }
  return result;
}

export class WebBookmarkCaptureProvider {
  async capture(input: CaptureInput, config: CaptureConfig): Promise<CaptureItem> {
    if (!input.url) throw new Error('web_bookmark capture requires a URL');

    const response = await fetch(input.url, {
      headers: { 'Range': `bytes=0-${MAX_HEAD_BYTES}` },
    });
    const html = await response.text();

    const og = extractOpenGraph(html);
    const twitter = extractTwitterCard(html);
    const htmlMeta: Partial<BookmarkData> = {
      title: extractHtmlTitle(html),
      description: extractMetaDescription(html),
      canonicalUrl: extractCanonicalUrl(html),
      favicon: extractFavicon(html, input.url),
      author: extractAuthor(html),
    };

    const bookmark = mergeBookmarkData(og, twitter, htmlMeta);
    const title = bookmark.title || 'Untitled Bookmark';

    const contentParts = [
      `# ${title}`,
      bookmark.description ? `\n${bookmark.description}` : '',
      bookmark.siteName ? `\nSite: ${bookmark.siteName}` : '',
      bookmark.author ? `\nAuthor: ${bookmark.author}` : '',
      bookmark.image ? `\nImage: ${bookmark.image}` : '',
      bookmark.favicon ? `\nFavicon: ${bookmark.favicon}` : '',
    ].filter(Boolean).join('');

    return {
      content: contentParts,
      sourceMetadata: {
        title,
        url: bookmark.canonicalUrl || input.url,
        capturedAt: new Date().toISOString(),
        contentType: 'application/x-bookmark',
        author: bookmark.author,
        tags: ['bookmark', bookmark.type || 'webpage'].filter(Boolean) as string[],
        source: 'web_bookmark',
      },
      rawData: config.options?.includeRawMeta ? bookmark : undefined,
    };
  }

  supports(input: CaptureInput): boolean {
    if (!input.url) return false;
    try {
      const parsed = new URL(input.url);
      return parsed.protocol === 'http:' || parsed.protocol === 'https:';
    } catch {
      return false;
    }
  }
}

export default WebBookmarkCaptureProvider;
